dts: stm32: uart clocks

erwango opened the pull request:
Provide rcc node to stm32 based socs.
Use rcc node to provide clock information on STM32 u(s)art nodes

erwango committed:
dts: arm: stm32: Add rcc node yaml descritpion

Provide rcc node yaml description for stm32 socs

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
dts: stm32: Add rcc node on stm32 socs

This commit provides rcc nodes to all stm32 based socs.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
dts: common: yaml: Update uart yaml with new properties

Update  uart yaml files to provide new properties:
-"interrupts-name"
-"pinctrl"

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

galak commented on dts/arm/st/clocks.h:
```diff
@@ -0,0 +1,19 @@
+#define STM32_CLOCK_BUS_AHB2    1
+#define STM32_CLOCK_BUS_APB1    2
+#define STM32_CLOCK_BUS_APB2    3
+#ifdef CONFIG_SOC_SERIES_STM32L4X
```
remove the #ifdef here.

galak commented on include/drivers/clock_control/stm32_clock_control.h:
```diff
@@ -19,10 +19,10 @@ enum {
 #endif
-	STM32_CLOCK_BUS_APB2,
 };
 
```
remove this enum, and just the defines from dts/arm/st/clocks.h.

erwango replied:
I'm able to #include <cloks.h>
I think it'd be better to #include <dts/arm/st/clocks.h>, but I can't get it work
Any idea how I could do that?

erwango replied:
So best I could do is #include <arm/st/clocks.h>
I'll change the plan and #include <dt_stm32_clocks.h>

galak commented on dts/arm/st/stm32f103Xb.dtsi:
```diff
@@ -32,6 +32,14 @@
+			clocks-controller;
+			#clocks-cells = <2>;
+			reg = <0x40021000 0x400>;
+			label = "STM32_CLOCK";
```
is the label here for code generation purposes?  If so how about STM32_CLK_RCC

erwango replied:
arbitrary choice, STM32_CLK_RCC is ok

galak reviewed (changes requested):
Some cleanup on enum / defines, but otherwise looks fine.

galak commented on dts/arm/yaml/st,stm32-rcc.yaml:
```diff
@@ -0,0 +1,28 @@
+cell_string: clocks
+
+"#cells":
+  - bus
```
is this correct, should we have:

- phandle
- bus
- bits

?

erwango replied:
I can have a try but I don't know if script will handle it.
It's a bit picky with clocks node (I'll fw you a mail about that)

erwango replied:
seems to work, let's go

ydamigos commented on dts/common/yaml/uart.yaml:
```diff
@@ -27,4 +27,14 @@ properties:
       description: Human readable string describing the device (used by Zephyr for API name)
       generation: define
+  - interrupts-name:
+      type: array
```
Do you mean "interrupt-names"?
The interrupt-names property consists of a list of strings so the "stringlist" should be used here.

erwango replied:
Indeed, thanks!

ydamigos commented:
According to linux device tree bindings for "st,stm32-rcc" (http://elixir.free-electrons.com/linux/v4.13-rc2/source/Documentation/devicetree/bindings/clock/st,stm32-rcc.txt):

> - #clock-cells: 2, device nodes should specify the clock in their "clocks"
>   property, containing a phandle to the clock device node, an index selecting
>   between gated clocks and other clocks and an index specifying the clock to
> use.

@erwango Do we plan to follow linux DT bindings for clocks or use something different for Zephyr?

erwango commented:
@ydamigos : Thanks for raising this question.
Long term plan is indeed to converge with Linux DT (not "converge to" Linux DT), but we don't want this to block zephyr short term objectives/progress, and hence I'd like to  be able to derogate when needed.

There are several reasons to that:
1-Using Linux bindings may require to either rework zephyr driver either to have binding not reflecting
zephyr implementation of being over complicated vs zephyr driver need. 
For instance on rcc node, Linux biding is the output of a complex driver (clock gating, ability to have clock per IP, ..) which is not in line with the current rcc driver on Zephyr. Having same bindings on Zephyr would be nice but would not reflect current Zephyr RCC driver.

2-"Linux time" is different from "Zephyr time", they have their own objectives and plan. Linux STM32 device tree is not frozen and evolutions are on-going. Given Linux upstream process, it can take time and may require several reworks (hence months), before we know on what we should align on.

3-Zephyr will have more stm32 dts files than Linux. If we want to have a dts collection of files coherent from {zephyr, Linux, ..} global point of view, Linux dts will have to accept choices that might not be
coherent from single Linux point of view.

This being said, if there are options to get things closer to Linux, or more coherent from global point of view, while keeping things simple in zephyr and not preventing us to progress at our expected pace, we should use them.

galak commented on dts/arm/Makefile:
```diff
@@ -1,4 +1,9 @@
 ifeq ($(CONFIG_HAS_DTS),y)
+
+ifdef CONFIG_SOC_FAMILY_STM32
+ZEPHYRINCLUDE += -I$(srctree)/dts/arm/st
```
Can we not just include via:

#include <dts/arm/st/dt_stm32_clocks.h>

erwango replied:
I had a try but didn't work. Maybe I was not lucid anymore but is there any subtlety to achieve this?
I guess #include "dts/arm/st/dt_stm32_clocks.h" would be more likely to work

galak commented on dts/arm/st/dt_stm32_clocks.h:
```diff
@@ -0,0 +1,17 @@
+/*
```
any reason not to just name this stm32_clocks.h (related to how we include it)

erwango replied:
Indeed I added dt_ to compensate the missing path

galak commented on dts/arm/st/dt_stm32_clocks.h:
```diff
@@ -0,0 +1,17 @@
+#define STM32_CLOCK_BUS_AHB2    1
+#define STM32_CLOCK_BUS_APB1    2
+#define STM32_CLOCK_BUS_APB2    3
+#define	STM32_CLOCK_BUS_APB1_2  4
```
some odd whitespace issue, guessing a tab after define

galak commented on include/drivers/clock_control/stm32_clock_control.h:
```diff
@@ -10,21 +10,11 @@
 #define _STM32_CLOCK_CONTROL_H_
 
 #include <clock_control.h>
+#include <dt_stm32_clocks.h>
```
lets include this via #include <dts/arm/st/dt_stm32_clocks.h>

galak reviewed (changes requested):
Some minor nits and how we access the includes

erwango commented:
@galak: Are you good with new version?

erwango committed:
dts: arm: Add clocks u(s)art on stm32 family

Add "clocks" property on u(s)arts nodes on stm32 socs
Add a dt clocks binding file and rework clock_control
header file include new device tree binding file.
include/dt-bindings folder is introduced as dt-bindings
placeholder

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>
